File: 581-Shortest-Unsorted-Continuous-Subarray.py
class Solution:
    def findUnsortedSubarray(self, nums: List[int]) -> int:
        start = 0
        while start < len(nums) - 1 and nums[start] <= nums[start + 1]:
            start += 1
        
        if start == len(nums) - 1:
            return 0
        
        end = len(nums) - 1
        while end > 0 and nums[end] >= nums[end - 1]:
            end -= 1
        
        mini = min(nums[start:end + 1])
        maxi = max(nums[start:end + 1])

        while start > 0 and nums[start - 1] > mini:
            start -= 1
        while end < len(nums) - 1 and nums[end + 1] < maxi:
            end += 1

        return end - start + 1

File: 581-Shortest-Unsorted-Continuous-Subarray.java
public class Solution {
    public int findUnsortedSubarray(int[] nums) {
        int start = 0;
        while (start < nums.length - 1 && nums[start] <= nums[start + 1]) {
            start++;
        }
        
        if (start == nums.length - 1) {
            return 0;
        }
        
        int end = nums.length - 1;
        while (end > 0 && nums[end] >= nums[end - 1]) {
            end--;
        }
        
        int mini = Integer.MAX_VALUE, maxi = Integer.MIN_VALUE;
        for (int i = start; i <= end; i++) {
            mini = Math.min(mini, nums[i]);
            maxi = Math.max(maxi, nums[i]);
        }
        
        while (start > 0 && nums[start - 1] > mini) {
            start--;
        }
        
        while (end < nums.length - 1 && nums[end + 1] < maxi) {
            end++;
        }
        
        return end - start + 1;
    }
}
